### PR TITLE
MRDL->MRTK Feature Bounding Box and App Bar

### PR DESCRIPTION
**Pull request by cre8ivepark**
mrtk_appbar_boundingbox

Overview
---
This example scene demonstrates how to make objects manipulatable using BoundingBoxRig script. The bounding box is a standard interface for manipulating objects in Windows Mixed Reality. Using gizmo on the corners and edges, you can scale or rotate the object. App Bar provides the button for entering/exiting 'Adjust' mode to enable/disable Bounding Box. For more information please see '[App Bar and Bounding Box' on Windows Dev Center.
](https://developer.microsoft.com/en-us/windows/mixed-reality/app_bar_and_bounding_box)

Demo video
---
https://gfycat.com/TameQuaintGnu

Supported devices
---
- HoloLens 
- Immersive headset 

Test Scene 
---
[BoundingBoxGizmoExample.unity
](https://github.com/johnppella/MixedRealityToolkit-Unity/blob/BoundingBoxWithGizmo/Assets/MixedRealityToolkit-Examples/UX/Scenes/BoundingBoxGizmoExample.unity)
mrtk_appbar_boundingbox_examplescene

Technical details 
---
You can find technical details on the README page: 
https://github.com/johnppella/MixedRealityToolkit-Unity/blob/BoundingBoxWithGizmo/Assets/MixedRealityToolkit-Examples/UX/Readme/README_BoundingBoxGizmoExample.md

Dependency
---
This example scene demonstrates the combination of BoundingBoxRig with TwoHandManipulatable script. 

Changes
---
- Fixes: #1497 #1522
#

### CODE REVIEW

**Comment by cre8ivepark** [comment]
Reopened a new pull request, with the merge of the latest Dev_Working_Branch.  Now dependancy is the MRTK/Standard shader #1743 which was used in the gizmo materials.
Thanks!

**Comment by StephenHodgson** [comment]
@cre8ivepark merge conflicts

**Comment by cre8ivepark** [comment]
Resolved conflicts. Merged Dev_Working_Branch including Cameron's MRTK/Shader related updates.

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBox.cs [hunk 1]
```diff
@@ -109,6 +109,26 @@ public Vector3 TargetBoundsLocalScale
...
+            get
+            {
+                return isVisible;
+            }
+            set
+            {
+                Transform scale = transform.GetChild(0);
+                Transform rig = scale.GetChild(0);
+                GameObject rigobject = rig.gameObject;
+                rigobject.gameObject.GetComponent<Renderer>().enabled = value;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBox.cs]: maybe we can cache the renderer when we set the target reference?

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBox.cs [hunk 2]
```diff
@@ -386,9 +406,10 @@ protected virtual void UpdateScaleTransform()
 
             transform.eulerAngles = rotation;
         }
-
         #endregion
 
+  
+
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBox.cs]: some of these line changes really aren't needed.

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+            {
+                OnInputUp(null);
+            }
+            eventData.Use();
+        }
+    }
+}
+
+
+
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs]: lots of empty lines here at the bottom.

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+            }
+        }
+        public void OnSourceDetected(SourceStateEventData eventData)
+        {
+        }
+        public void OnSourceLost(SourceStateEventData eventData)
+        {
+            if (eventData.SourceId == inputDownEventData.SourceId)
+            {
+                OnInputUp(null);
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs]: We probably shouldn't be calling this method like that.
Might be better if we had a private method that handed the logic

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+            Rig.FocusOnHandle(this.gameObject);
+            eventData.Use();
+        }
+        public void OnInputUp(InputEventData eventData)
+        {
+            inputDownEventData = null;
+
+            if (this.AffineType == TransformType.Scale)
+            {
+                this.gameObject.GetComponent<Renderer>().material = Rig.ScaleHandleMaterial;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs]: memory leak

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+            }
+            else
+            {
+                this.gameObject.GetComponent<Renderer>().material = Rig.RotateHandleMaterial;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs]: memory leak

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+            initialHandPosition     = GetHandPosition(eventData.SourceId);
+            initialScale            = transformToAffect.localScale;
+            initialPosition         = transformToAffect.position;
+            initialOrientation      = transformToAffect.rotation.eulerAngles;
+            initialHandOrientation  = GetHandOrientation(eventData.SourceId);
+            initialScaleOrigin      = transformToAffect.position - this.transform.position;
+
+            MixedRealityToolkit.InputModule.InputManager.Instance.PushModalInputHandler(gameObject);
+
+            this.gameObject.GetComponent<Renderer>().material = Rig.InteractingMaterial;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoHandle.cs]: memory leak

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+        private Material scaleHandleMaterial;
+        [SerializeField]
+        private Material rotateHandleMaterial;
+        [SerializeField]
+        private Material interactingMaterial;
+
+        [Header("Preset Components")]
+        [SerializeField]
+        [Tooltip("To visualize the object bounding box, drop the MixedRealityToolkit/UX/Prefabs/BoundingBoxes/BoundingBoxBasic.prefab here.")]
+        public BoundingBox BoundingBoxPrefab;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs]: if this is serialized, then it should also be private.
if access outside the class is needed, then please create a property and use the field as it's backing.

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+            ClearHandles();
+            UpdateCornerHandles();
+            UpdateRotateHandles();
+            ParentHandles();
+            UpdateHandles();
+        }
+
+        private void UpdateCornerHandles()
+        {
+            handleCentroids = handleCentroids ?? GetBounds();
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs]: not sure ?? works the way you expect for `UnityEngine.Object` types.
Specifically checking `!=null` is best practice here.

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+
+            if (cornerHandles == null)
+            {
+                cornerHandles = new GameObject[handleCentroids.Count];
+                for (int i = 0; i < cornerHandles.Length; ++i)
+                {
+                    cornerHandles[i] = GameObject.CreatePrimitive(PrimitiveType.Cube);
+                    cornerHandles[i].GetComponent<Renderer>().material = scaleHandleMaterial;
+                    cornerHandles[i].transform.localScale = scaleHandleSize;
+                    cornerHandles[i].AddComponent<BoundingBoxGizmoHandle>();
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs]: cache the component here and use it everywhere below.
```
var handle = cornerHandles[i].AddComponent<BoundingBoxGizmoHandle>();
handle.Rig = this;
etc...
```

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+            for (int i = 0; i < cornerHandles.Length; ++i)
+            {
+                cornerHandles[i].transform.position = handleCentroids[i];
+                cornerHandles[i].transform.localRotation = objectToBound.transform.rotation;
+            }
+        }
+
+        private void UpdateRotateHandles()
+        {
+            handleCentroids = handleCentroids ?? GetBounds();
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs]: better to check null

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+            if (rotateHandles == null)
+            {
+                rotateHandles = new GameObject[12];
+
+                for (int i = 0; i < rotateHandles.Length; ++i)
+                {
+                    rotateHandles[i] = GameObject.CreatePrimitive(PrimitiveType.Sphere);
+                    rotateHandles[i].GetComponent<Renderer>().material = rotateHandleMaterial;
+                    rotateHandles[i].transform.localScale = rotateHandleSize;
+                    rotateHandles[i].AddComponent<BoundingBoxGizmoHandle>();
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs]: cache and use the handle component.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs]: You could even make a cache for each handle and re-use it outside the for loop below so you don't have to use get component on each again.

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+        private List<Vector3> GetBounds()
+        {
+            if (objectToBound != null)
+            {
+                List<Vector3> bounds = new List<Vector3>();
+                LayerMask mask = new LayerMask();
+
+                GameObject clone = GameObject.Instantiate(boxInstance.gameObject);
+                clone.transform.localRotation = Quaternion.identity;
+                clone.transform.position = new Vector3(0, 0, 0);
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs]: use `clone.transform.position = Vector3.zero`;

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs (hunk 9 above)
Comment (DDReaper) [reply Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs]: Due to the current state of Unity, we do not advise using the null coalescing operator.  Please revert to the old if == null check.

Comment (johnppella) [reply Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxRig.cs]: will do

**Comment by johnppella** [comment]
changes made: 
convention fixes
memleak fix
simplifications
as per requested changes

**Comment by johnppella** [comment]
got rid of extra lines in Boundingbox.cs and boundingboxgizmohandle.cs

**Comment by cre8ivepark** [comment]
Updated landing page README to include Bounding Box. 
https://github.com/johnppella/MixedRealityToolkit-Unity/tree/Feature_UX_BoundingBoxAppBar

**Comment by DDReaper** [comment]
I would like to see this also ported to the next Stabilisation branch if possible, provided it's not dependant on experimental features.

**Pull request merged**